README
POE-Component-Client-SMTP

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it get an idea of the modules uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.

INSTALLATION

To install this module, run the following commands:

    perl Build.PL
    ./Build
    ./Build test
    ./Build install


COPYRIGHT AND LICENCE

Copyright (C) 2005 - 2008 George Nistorica

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
Revision history for POE-Component-Client-SMTP
0.21  February 2009
    * bug #39345 [Provide a consolidated error message]
      Reported by: David Golden
    * pococlsmtp was sending the error event to the caller
      multiple times per SMTP transaction if it had things in the
      queue to send to the server
      Reported by: I lost the email, please send me the mail again!
    * escaping single dot in the message body (slurped file or
      message body as provided by the user)
      Reported by: I lost the email, please send me the mail again!
    * bug #39290 Document that From and To must be bare email addresses
      documentation specifing the fact that To/From arguments
      need to hold bare email addresses not RFC2822 email address
      Reported by: David Golden
    * Fix the bug when sending file content and receiving a server
      answer

0.20	May 2008
    * Fixed NetBSD failed test

0.19	May 2008
    * Use POE::Filter::Transparent::SMTP for sending messages

0.18    Aug 2007
    * Generate a Makefile.PL also because of test failures generated by smokers
    that don't have Module::Build installed

0.17    May 2007
    * Added support for TransactionLog, logging the SMTP session and posting it
    back to the caller session

0.16    May 2007
    * Added support for reading email body from files, via the parameters:
        MessageFile and FileHandle

0.15    February-16-2007
    * Added ESMTP support for PLAIN AUTH (request from David Janssens)

0.14	January-25-2007
    * Fixed bug 24541 ( http://rt.cpan.org/Ticket/Display.html?id=24541 )
    reported by Thomas Klausner ( DOMM )
    "In line 182 of PoCo::Client::SMTP.pm, you check the return value via a
    regular expression. But if the return value indicates an error
    (according to RFC 2821 Section 4.2.1 if the retrun value starts with
    5xx), you still continue to post the data. And in the end, the
    SMTP_Failure event is never called."

0.13    February-06-2006
    * Fixed a bug concerning the return values in case of error sending an email
        This means that ARG1 sent back when SMTP_Failure event occurs, having a
        key like OE::Wheel::* will contain a ref to an array containing the
        error received by the underlying wheel
        ( Thanks to Sergey Skvortsov for pointing me that out )
    * Added test file for checking the return when POE::Wheel::SocketFactory
        fails

0.12    November-17-2005
    * Added BindAddress and BindPort attributes that are passed to
POE::Wheel::SocketFactory when creating the socket.

0.11    November-11-2005
    * refactored the module to use Objects
    * redone the API
    * added examples
    * added BCC/CC support
    * added context support (Mike Schroeder's suggestion)
    * removed Alias as a mandatory parameter (patch from BinGOs)

0.06 - 0.10 Development or inexistent versions

0.05    June-23-2005
    * added smtp_bind_address parameter to ->send();
    it is passed to SocketFactory's constructor as BindAddress

0.04    June-20-2005
    * added a new parameter for the send() method, smtp_body instead
    of smtp_data
    * smtp_data is mutually exclusive with any parameter that may
    alter the message's header; this means that you either send a
    simple message assembled by POE::Component::Client::SMTP using
    smtp_body or you construct the entire message elsewhere and
    pass it to the smtp_data parameter
    * added smtp_ehlo so that the user may control what PoCoClSMTP
    sends in response of server greeting

0.03    June-06-2005/00:00
        * removed t/07-isa.t

0.02    June-02-2005/00:00
    * Output in debug mode is a little bit more readable
        * Added SMTP "Timeout" facility
        * Added tests for the module
        * "cleaned" up the module inside, meaning that most of the SMTP server
        codes have their own handler

0.01    May-23-2005/17:00
        First version, released on an unsuspecting world.


t/10-send.pl
#!/usr/bin/perl -w
use strict;

use lib '../lib';

use IO::Socket::INET;

use POE qw(Wheel::ListenAccept);

use POE::Component::Client::SMTP;

use Test::More tests => 1;

my $success   = 0;
my $mail_body =
  "Ce faci bro' ? iar ai uitat de mine?\n" . "Hai! \r\n\la munca!!!!\n";

my $sender    = 'george@localhost';
my $recipient = 'george@localhost';
my $server    = 'localhost';
my $port      = 25;

POE::Session->create(
    inline_states => {
        _start       => \&start,
        send_mail    => \&send_mail,
        smtp_success => \&smtp_success,
        smtp_error   => \&smtp_error,
        _stop        => \&stop,
    },
    heap => { smtp_body => \$mail_body, },
);

POE::Kernel->run();

sub start {
    $_[KERNEL]->yield("send_mail");
}

sub send_mail {

    my $ref_to_data;
    my $data;

    $data        = "test test\nsdasdas\n";
    $ref_to_data = \$data;

    POE::Component::Client::SMTP->send(
        alias          => 'smtp_client',
        smtp_server    => $server,
        smtp_port      => $port,
        smtp_sender    => $sender,
        smtp_recipient => $recipient,
        to             => "Cos",
        from           => "Crony",
        subject        => "brrr, iar stai degeaba?!",
        smtp_body      => $_[HEAP]->{'smtp_body'},
        smtp_timeout   => 1,
        debug          => 2,

        SMTPSendSuccessEvent => "smtp_success",
        SMTPSendFailureEvent => "smtp_error",
    );
}

sub smtp_success {
    warn "success";
    $success = 1;
}

sub smtp_error {
    warn "eroare";
    my ( $err_type, $operation, $errnum, $errstr ) = @_[ ARG0 .. ARG4 ];
    warn "er: $err_type";
    my $heap = $_[HEAP];
    if ( $err_type == 2 ) {
        $success = 1;
        warn
"Could not connect to: $server:$port, error: $errnum, $errstr while $operation";
    }
    else {
        warn
"Could not connect to: $server:$port, error: $errnum, $errstr while $operation";
        $success = 0;
    }
}

sub stop {
    is( $success, 1,
        "Sending mail:from $sender, to $recipient, trough $server" );
}


eg/test_template.pl
#!/usr/bin/env perl 

# Copyright (c) 2005 - 2008 George Nistorica
# All rights reserved.
# This file is part of POE::Component::Client::SMTP
# POE::Component::Client::SMTP is free software; you can redistribute it and/or
# modify it under the same terms as Perl itself.  See the LICENSE
# file that comes with this distribution for more details.

use strict;
use warnings;

use lib '../lib';
use Test::More tests => 6;    # including use_ok
use Data::Dumper;
use Carp;

BEGIN { use_ok("IO::Socket::INET"); }
BEGIN { use_ok("POE"); }
BEGIN { use_ok("POE::Wheel::ListenAccept"); }
BEGIN { use_ok("POE::Component::Server::TCP"); }
BEGIN { use_ok("POE::Component::Client::SMTP"); }

my $test = 'undef';

my $smtp_message;
my @recipients;
my $from;
my $debug = 0;

$smtp_message = create_smtp_message();
@recipients   = qw(
);
$from = '';

##### SMTP server vars
my $port                  = 25252;
my $EOL                   = "\015\012";
my @smtp_server_responses = (
    "220 localhost ESMTP POE::Component::Client::SMTP Test Server",
    "250-localhost$EOL"
      . "250-PIPELINING$EOL"
      . "250-SIZE 250000000$EOL"
      . "250-VRFY$EOL"
      . "250-ETRN$EOL"
      . "250 8BITMIME",
    "250 Ok",                                 # mail from
    "250 Ok",                                 # rcpt to:
    "354 End data with <CR><LF>.<CR><LF>",    # data
    "250 Ok: queued as 549B14484F",           # end data
    "221 Bye",                                # quit
);

POE::Component::Server::TCP->new(
    Port                  => $port,
    Address               => "localhost",
    Domain                => AF_INET,
    Alias                 => "smtp_server",
    Error                 => \&error_handler,               # Optional.
    ClientInput           => \&handle_client_input,         # Required.
    ClientConnected       => \&handle_client_connect,       # Optional.
    ClientDisconnected    => \&handle_client_disconnect,    # Optional.
    ClientError           => \&handle_client_error,         # Optional.
    ClientFlushed         => \&handle_client_flush,         # Optional.
    ClientFilter          => "POE::Filter::Line",           # Optional.
    ClientInputFilter     => "POE::Filter::Line",           # Optional.
    ClientOutputFilter    => "POE::Filter::Line",           # Optional.
    ClientShutdownOnError => 1,                             #
);

POE::Session->create(
    inline_states => {
        _start             => \&start_session,
        _stop              => \&stop_session,
        send_mail          => \&spawn_pococlsmt,
        pococlsmtp_success => \&smtp_send_success,
        pococlsmtp_failure => \&smtp_send_failure,
    },
);

POE::Kernel->run();

is( $test, 1, "TEMPLATE" );
diag("TEMPLATE");

sub start_session {
    $_[KERNEL]->yield("send_mail");
}

sub spawn_pococlsmt {
    POE::Component::Client::SMTP->send(
        From         => $from,
        To           => \@recipients,
        SMTP_Success => 'pococlsmtp_success',
        SMTP_Failure => 'pococlsmtp_failure',
        Server       => 'localhost',
        Port         => $port,
        Body         => $smtp_message,
        Context      => "test context",
    );
}

sub stop_session {

    # stop server
    $_[KERNEL]->call( smtp_server => "shutdown" );
}

sub smtp_send_success {
    my ( $arg0, $arg1 ) = @_[ ARG0, ARG1 ];
    print "ARG0, ", Dumper($arg0), "\nARG1, ", Dumper($arg1) if $debug;
    $test = 1;
}

sub smtp_send_failure {
    my ( $arg0, $arg1 ) = @_[ ARG0, ARG1 ];
    print "ARG0, ", Dumper($arg0), "\nARG1, ", Dumper($arg1) if $debug;
    $test = 0;
}

sub create_smtp_message {
    my $body = <<EOB;
To: George Nistorica <george\@localhost>
Bcc: George Nistorica <george\@localhost>
CC: Alter Ego <root\@localhost>
From: Charlie Root <root\@localhost>
Subject: Email test

Sent with $POE::Component::Client::SMTP::VERSION
EOB

    return $body;
}

sub error_handler {
    carp "Something nasty happened";
    exit 100;
}

sub handle_client_input {
    my ( $heap, $input ) = @_[ HEAP, ARG0 ];

    if ( $input =~ /^(helo|ehlo|mail from:|rcpt to:|data|\.|quit)/i ) {
        print "$input\n" if $debug;
        $heap->{'client'}->put( shift @smtp_server_responses );
    }
}

sub handle_client_connect {
    $_[HEAP]->{'client'}->put( shift @smtp_server_responses );
}

sub handle_client_disconnect {
}

sub handle_client_error {
}

sub handle_client_flush {
}



Build.PL
# Copyright (c) 2005 - 2009 George Nistorica
# All rights reserved.
# This file is part of POE::Component::Client::SMTP
# POE::Component::Client::SMTP free software; you can redistribute it and/or
# modify it under the same terms as Perl itself.  See the LICENSE
# file that comes with this distribution for more details.

# 	$Id: Build.PL,v 1.8 2008/05/13 14:02:14 UltraDM Exp $

use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    q{module_name}       => q{POE::Component::Client::SMTP},
    q{license}           => q{perl},
    q{dist_author}       => q{George Nistorica <ultadm@cpan.org>},
    q{dist_version_from} => q{lib/POE/Component/Client/SMTP.pm},
    q{requires}          => {
        q{perl}                           => 0,
        q{Test::More}                     => 0,
        q{Carp}                           => 0,
        q{Socket}                         => 0,
        q{Data::Dumper}                   => 0,
        q{POE}                            => q{0.26},
        q{POE::Wheel::SocketFactory}      => 0,
        q{POE::Wheel::ReadWrite}          => 0,
        q{POE::Filter::Line}              => 0,
        q{POE::Filter::Stream}            => 0,
        q{Symbol}                         => 0,
        q{POE::Filter::Transparent::SMTP} => q{0.2},
    },
    q{add_to_cleanup}     => [q{POE-Component-Client-SMTP-*}],
    q{create_makefile_pl} => q{traditional},
);

$builder->create_build_script();


eg/send_email.pl
#!/usr/bin/perl -w
use strict;

# Copyright (c) 2005 George Nistorica
# All rights reserved.
# This file is part of POE::Component::Client::SMTP
# POE::Component::Client::SMTP is free software; you can redistribute it and/or
# modify it under the same terms as Perl itself.  See the LICENSE
# file that comes with this distribution for more details.

my $sender      = 'replace@with.email.address.net';
my $recipient   = 'replace@with.email.address.net';
my $smtp_server = 'your.relay.mail.server.net';
my $smtp_port   = 25;

# use the library from the kit
# remove the line below if you're using the system wide installed
# PoCoClSMTP
use lib '../lib';

use Data::Dumper;    # I always include this ;-)
use Email::MIME::Creator;
use IO::All;

use POE;
use POE::Component::Client::SMTP;

# main()
POE::Session->create(
    inline_states => {
        _start            => \&start_main_session,
        send_mail         => \&send_mail_from_main_session,
        send_mail_success => \&send_mail_success,
        send_mail_failure => \&send_mail_failure,
        _stop             => \&stop_main_session,
    }
);

POE::Kernel->run();

# done

sub start_main_session {

    #fire the things up
    $_[KERNEL]->yield("send_mail");
}

sub send_mail_from_main_session {
    my $email = create_message();

    # Note that you are prohibited by RFC to send bare LF characters in e-mail
    # messages; consult:
    # http://cr.yp.to/docs/smtplf.html
    $email =~ s/\n/\r\n/g;

    POE::Component::Client::SMTP->send(
        From         => $sender,
        To           => $recipient,
        Server       => $smtp_server,
        Port         => $smtp_port,
        Body         => $email,
        SMTP_Success => 'send_mail_success',
        SMTP_Failure => 'send_mail_failure',
    );
}

sub send_mail_success {
    print "Success\n";
}

sub send_mail_failure {
    my $fail = $_[ARG1];
    print Dumper($fail);
    print "Failure\n";
}

sub stop_main_session {
    print "End ...\n";
}

# Email Creation Part
# rather lame email creation.
# You may use any method that suits you (I usually create the messages by hand
# ;-) )
sub create_message {
    my $attachment_file = "text_mail_attachment.txt";

    my $email;
    my @parts;

    @parts = (
        Email::MIME->create(
            attributes => {
                filename     => "text.txt",
                content_type => "text/plain",
                encoding     => "quoted-printable",
                name         => "Example attachment",
            },
            body => io($attachment_file)->all,
        ),
        Email::MIME->create(
            attributes => {
                content_type => "text/plain",
                disposition  => "attachment",
                charser      => "US-ASCII",
            },
            body => "Howdy!",
        ),
    );

    $email = Email::MIME->create(
        header => [
            From => $sender,
            To   => $recipient,
        ],
        parts => [@parts],
    );

    # return the message
    return $email->as_string;
}



eg/send_email_auth.pl
#!/usr/bin/perl -w
use strict;

# Copyright (c) 2005 - 2007 George Nistorica
# All rights reserved.
# This file is part of POE::Component::Client::SMTP
# POE::Component::Client::SMTP is free software; you can redistribute it and/or
# modify it under the same terms as Perl itself.  See the LICENSE
# file that comes with this distribution for more details.

my $sender      = 'george@localhost';
my $recipient   = 'george@localhost';
my $smtp_server = 'localhost';
my $user        = 'george', my $pass = 'abracadabra', my $smtp_port = 25;

# use the library from the kit
# remove the line below if you're using the system wide installed
# PoCoClSMTP
use lib '../lib';

use Data::Dumper;    # I always include this ;-)
use Email::MIME::Creator;
use IO::All;

use POE;
use POE::Component::Client::SMTP;

# main()

print "V: ", $POE::Component::Client::SMTP::VERSION, "\n";
POE::Session->create(
    inline_states => {
        _start            => \&start_main_session,
        send_mail         => \&send_mail_from_main_session,
        send_mail_success => \&send_mail_success,
        send_mail_failure => \&send_mail_failure,
        _stop             => \&stop_main_session,
    }
);

POE::Kernel->run();

# done

sub start_main_session {

    #fire the things up
    $_[KERNEL]->yield("send_mail");
}

# 9qi<|'K3,8fJQpt!#z%S6AA9j8'2
sub send_mail_from_main_session {
    my $email = create_message();

    # Note that you are prohibited by RFC to send bare LF characters in e-mail
    # messages; consult:
    # http://cr.yp.to/docs/smtplf.html
    $email =~ s/\n/\r\n/g;

    POE::Component::Client::SMTP->send(
        From   => $sender,
        To     => $recipient,
        Server => $smtp_server,
        Auth   => {
            mechanism => 'PLAIN',
            user      => $user,
            pass      => $pass,
        },
        Port         => $smtp_port,
        Body         => $email,
        SMTP_Success => 'send_mail_success',
        SMTP_Failure => 'send_mail_failure',
        Debug        => 1,
    );
}

sub send_mail_success {
    print "Success\n";
}

sub send_mail_failure {
    my $fail = $_[ARG1];
    print Dumper($fail);
    print "Failure\n";
}

sub stop_main_session {
    print "End ...\n";
}

# Email Creation Part
# rather lame email creation.
# You may use any method that suits you (I usually create the messages by hand
# ;-) )
sub create_message {
    my $attachment_file = "text_mail_attachment.txt";

    my $email;
    my @parts;

    @parts = (
        Email::MIME->create(
            attributes => {
                filename     => "text.txt",
                content_type => "text/plain",
                encoding     => "quoted-printable",
                name         => "Example attachment",
            },
            body => io($attachment_file)->all,
        ),
        Email::MIME->create(
            attributes => {
                content_type => "text/plain",
                disposition  => "attachment",
                charser      => "US-ASCII",
            },
            body => "Howdy!",
        ),
    );

    $email = Email::MIME->create(
        header => [
            From => $sender,
            To   => $recipient,
        ],
        parts => [@parts],
    );

    # return the message
    return $email->as_string;
}



Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.03
use ExtUtils::MakeMaker;
WriteMakefile
(
          'NAME' => 'POE::Component::Client::SMTP',
          'VERSION_FROM' => 'lib/POE/Component/Client/SMTP.pm',
          'PREREQ_PM' => {
                           'Carp' => '0',
                           'Data::Dumper' => '0',
                           'POE' => '0.26',
                           'POE::Filter::Line' => '0',
                           'POE::Filter::Stream' => '0',
                           'POE::Filter::Transparent::SMTP' => '0',
                           'POE::Wheel::ReadWrite' => '0',
                           'POE::Wheel::SocketFactory' => '0',
                           'Socket' => '0',
                           'Symbol' => '0',
                           'Test::More' => '0'
                         },
          'INSTALLDIRS' => 'site',
          'EXE_FILES' => [],
          'PL_FILES' => {}
        )
;
